filer/models/filemodels.py

```diff
@@ -469,8 +474,8 @@ def restore(self):
             self.folder = filer.models.Folder.objects.get(id=self.folder_id)
 
         old_location, new_location = self.file.name, None
-        destination = self._generate_valid_location_for_restore()
-
+        self._generate_valid_name_for_restore()
+        destination = self.file.field.upload_to(self, self.upload_to_name)
```

Comment:
The old generate function did 2 things, mutate the state and return the generated destination.
That was hard to understand so that is why I split them. I hope that when these 2 lines are read it's clear that the generate method mutates the state and not hidden.
A lot of functions here should be simplified to only compute or to only mutate, but in this case the `if self.name in ('', None)` should be moved in a state mutate function and use the result from a generate function. 
Do you want to do it like this ?